File: codex_token_trace/parser.py
from __future__ import annotations

import glob
import json
import shlex
from pathlib import Path
from typing import Any, Iterable

from .models import Event, Interval, SessionAnalysis, TokenEvent, ToolCall, Usage


CALL_TYPES = {"function_call", "custom_tool_call", "web_search_call"}
OUTPUT_TYPES = {"function_call_output", "custom_tool_call_output", "web_search_end"}


def _payload_type(obj: dict[str, Any]) -> str | None:
    payload = obj.get("payload")
    if isinstance(payload, dict):
        return payload.get("type")
    return None


def _usage(data: dict[str, Any] | None) -> Usage:
    if not isinstance(data, dict):
        return Usage()
    return Usage(
        input_tokens=int(data.get("input_tokens") or 0),
        cached_input_tokens=int(data.get("cached_input_tokens") or 0),
        output_tokens=int(data.get("output_tokens") or 0),
        reasoning_output_tokens=int(data.get("reasoning_output_tokens") or 0),
        total_tokens=int(data.get("total_tokens") or 0),
    )


def _json_len(value: Any) -> int:
    if value is None:
        return 0
    if isinstance(value, str):
        return len(value)
    return len(json.dumps(value, ensure_ascii=False, separators=(",", ":")))


def _decode_command(arguments: Any) -> tuple[str, int]:
    if arguments is None:
        return "", 0
    if isinstance(arguments, str):
        arg_chars = len(arguments)
        try:
            decoded = json.loads(arguments)
        except json.JSONDecodeError:
            return arguments, arg_chars
    else:
        decoded = arguments
        arg_chars = _json_len(arguments)

    if isinstance(decoded, dict):
        cmd = decoded.get("cmd") or decoded.get("command") or decoded.get("input") or decoded.get("path") or ""
        if isinstance(cmd, list):
            return shlex.join(str(part) for part in cmd), arg_chars
        return str(cmd), arg_chars
    return str(decoded), arg_chars


def _call_id(payload: dict[str, Any]) -> str | None:
    value = payload.get("call_id") or payload.get("id")
    return str(value) if value else None


def _tool_name(payload: dict[str, Any], fallback: str | None) -> str | None:
    value = payload.get("name") or payload.get("tool_name") or payload.get("tool")
    return str(value) if value else fallback


def _output_chars(payload: dict[str, Any]) -> int:
    for key in ("output", "result", "content", "text"):
        if key in payload:
            return _json_len(payload.get(key))
    return _json_len(payload)


def parse_session_file(path: Path) -> SessionAnalysis:
    events: list[Event] = []
    token_events: list[TokenEvent] = []
    tool_by_id: dict[str, ToolCall] = {}
    session_meta_chars_total = 0
    session_meta_chars_max = 0
    session_meta_count = 0
    previous_total_tokens: int | None = None

    with path.open("r", encoding="utf-8") as fh:
        for line_no, line in enumerate(fh, start=1):
            line = line.rstrip("\n")
            if not line:
                continue
            try:
                obj = json.loads(line)
            except json.JSONDecodeError:
                continue

            raw_chars = len(line)
            top_type = obj.get("type")
            ptype = _payload_type(obj)
            event = Event(
                line_no=line_no,
                timestamp=obj.get("timestamp"),
                top_type=top_type,
                payload_type=ptype,
                raw_chars=raw_chars,
                raw=obj,
            )
            events.append(event)

            if top_type == "session_meta":
                session_meta_count += 1
                session_meta_chars_total += raw_chars
                session_meta_chars_max = max(session_meta_chars_max, raw_chars)

            payload = obj.get("payload") if isinstance(obj.get("payload"), dict) else {}
            if ptype in CALL_TYPES:
                call_id = _call_id(payload)
                if call_id:
                    command, arg_chars = _decode_command(payload.get("arguments") or payload.get("input"))
                    tool_by_id[call_id] = ToolCall(
                        call_id=call_id,
                        timestamp=obj.get("timestamp"),
                        name=_tool_name(payload, ptype),
                        command=command,
                        arguments_chars=arg_chars,
                        call_type=ptype or "tool_call",
                    )
            elif ptype in OUTPUT_TYPES:
                call_id = _call_id(payload)
                if call_id:
                    tool = tool_by_id.get(call_id)
                    if tool is None:
                        tool = ToolCall(
                            call_id=call_id,
                            timestamp=obj.get("timestamp"),
                            name=_tool_name(payload, ptype),
                            command="",
                            arguments_chars=0,
                            call_type=ptype or "tool_output",
                        )
                        tool_by_id[call_id] = tool
                    tool.output_chars += _output_chars(payload)
                    tool.output_events += 1
            elif ptype == "token_count":
                info = payload.get("info")
                if isinstance(info, dict):
                    total = _usage(info.get("total_token_usage"))
                    duplicate = previous_total_tokens == total.total_tokens
                    previous_total_tokens = total.total_tokens
                    token = TokenEvent(
                        line_no=line_no,
                        timestamp=obj.get("timestamp"),
                        top_type=top_type,
                        payload_type=ptype,
                        raw_chars=raw_chars,
                        raw=obj,
                        last=_usage(info.get("last_token_usage")),
                        total=total,
                        context_window=info.get("model_context_window"),
                        duplicate_total=duplicate,
                    )
                    token_events.append(token)

    tools = sorted(tool_by_id.values(), key=lambda c: c.output_chars, reverse=True)
    intervals = _build_intervals(events, token_events, tool_by_id)
    unique_token_events = [e for e in token_events if not e.duplicate_total]

    return SessionAnalysis(
        path=path,
        events=events,
        token_events=token_events,
        unique_token_events=unique_token_events,
        tools=tools,
        intervals=intervals,
        session_meta_chars_total=session_meta_chars_total,
        session_meta_chars_max=session_meta_chars_max,
        session_meta_count=session_meta_count,
    )


def _build_intervals(events: list[Event], token_events: list[TokenEvent], tool_by_id: dict[str, ToolCall]) -> list[Interval]:
    by_line = {event.line_no: event for event in events}
    intervals: list[Interval] = []
    previous_line: int | None = None

    for token in token_events:
        if token.duplicate_total:
            previous_line = token.line_no
            continue
        start = (previous_line or 0) + 1
        end = token.line_no - 1
        between = [by_line[i] for i in range(start, end + 1) if i in by_line]
        raw_chars_by_type: dict[str, int] = {}
        count_by_type: dict[str, int] = {}
        call_ids: set[str] = set()

        for event in between:
            key = event.type_key
            raw_chars_by_type[key] = raw_chars_by_type.get(key, 0) + event.raw_chars
            count_by_type[key] = count_by_type.get(key, 0) + 1
            payload = event.raw.get("payload")
            if isinstance(payload, dict):
                call_id = _call_id(payload)
                if call_id:
                    call_ids.add(call_id)

        tools = [tool_by_id[cid] for cid in call_ids if cid in tool_by_id]
        tools.sort(key=lambda c: c.output_chars, reverse=True)
        intervals.append(
            Interval(
                previous_token_line=previous_line,
                token_event=token,
                event_count=len(between),
                raw_chars_by_type=raw_chars_by_type,
                count_by_type=count_by_type,
                tools=tools,
            )
        )
        previous_line = token.line_no

    return intervals


def discover_files(paths: Iterable[str]) -> list[Path]:
    files: list[Path] = []
    for raw in paths:
        p = Path(raw).expanduser()
        if any(ch in raw for ch in "*?["):
            files.extend(Path(match) for match in glob.glob(str(p), recursive=True))
        elif p.is_dir():
            files.extend(p.rglob("*.jsonl"))
        elif p.is_file():
            files.append(p)
    return sorted(set(files))

File: src/parser.ts
import { readdirSync, readFileSync, statSync } from "node:fs";
import { basename, extname, join, resolve } from "node:path";
import type { EventRecord, Interval, JsonObject, JsonValue, SessionAnalysis, TokenEvent, ToolCall, Usage } from "./models.js";
import { emptyUsage, eventTypeKey } from "./models.js";

const CALL_TYPES = new Set(["function_call", "custom_tool_call", "web_search_call"]);
const OUTPUT_TYPES = new Set(["function_call_output", "custom_tool_call_output", "web_search_end"]);

type MutableToolCall = {
  callId: string;
  timestamp: string | null;
  name: string | null;
  command: string;
  argumentsChars: number;
  callType: string;
  inputPreview: string;
  inputChars: number;
  outputChars: number;
  outputEvents: number;
};

const isObject = (value: JsonValue | undefined): value is JsonObject =>
  typeof value === "object" && value !== null && !Array.isArray(value);

const asString = (value: JsonValue | undefined): string | null => {
  if (typeof value === "string") {
    return value;
  }
  if (typeof value === "number" || typeof value === "boolean") {
    return String(value);
  }
  return null;
};

const payloadType = (row: JsonObject): string | null => {
  const payload = row.payload;
  if (!isObject(payload)) {
    return null;
  }
  return asString(payload.type);
};

const jsonLength = (value: JsonValue | undefined): number => {
  if (value === undefined || value === null) {
    return 0;
  }
  if (typeof value === "string") {
    return value.length;
  }
  return JSON.stringify(value).length;
};

const usageFrom = (value: JsonValue | undefined): Usage => {
  if (!isObject(value)) {
    return emptyUsage();
  }
  return {
    inputTokens: Number(value.input_tokens ?? 0),
    cachedInputTokens: Number(value.cached_input_tokens ?? 0),
    outputTokens: Number(value.output_tokens ?? 0),
    reasoningOutputTokens: Number(value.reasoning_output_tokens ?? 0),
    totalTokens: Number(value.total_tokens ?? 0),
  };
};

const parseJsonObject = (text: string): JsonObject | null => {
  try {
    const parsed = JSON.parse(text) as JsonValue;
    return isObject(parsed) ? parsed : null;
  } catch {
    return null;
  }
};

const decodeInput = (value: JsonValue | undefined): { label: string; chars: number; preview: string } => {
  if (value === undefined || value === null) {
    return { label: "", chars: 0, preview: "" };
  }

  const chars = jsonLength(value);
  const decoded = typeof value === "string" ? (parseJsonObject(value) ?? value) : value;

  if (typeof decoded === "string") {
    return { label: decoded, chars, preview: decoded };
  }
  if (Array.isArray(decoded)) {
    const label = decoded.map((item) => String(item)).join(" ");
    return { label, chars, preview: label };
  }
  if (isObject(decoded)) {
    const candidate =
      decoded.cmd ?? decoded.command ?? decoded.input ?? decoded.path ?? decoded.filepath ?? decoded.url;
    if (Array.isArray(candidate)) {
      const label = candidate.map((item) => String(item)).join(" ");
      return { label, chars, preview: label };
    }
    if (candidate !== undefined && candidate !== null) {
      const label = typeof candidate === "string" ? candidate : JSON.stringify(candidate);
      return { label, chars, preview: label };
    }
    const label = JSON.stringify(decoded);
    return { label, chars, preview: label };
  }
  const label = String(decoded);
  return { label, chars, preview: label };
};

const callIdOf = (payload: JsonObject): string | null =>
  asString(payload.call_id) ?? asString(payload.id);

const toolNameOf = (payload: JsonObject, fallback: string | null): string | null =>
  asString(payload.name) ?? asString(payload.tool_name) ?? asString(payload.tool) ?? fallback;

const outputCharsOf = (payload: JsonObject): number => {
  for (const key of ["output", "result", "content", "text"] as const) {
    if (key in payload) {
      return jsonLength(payload[key]);
    }
  }
  return jsonLength(payload);
};

const toToolCall = (tool: MutableToolCall): ToolCall => ({ ...tool });

const updateOutput = (tool: MutableToolCall, outputChars: number): void => {
  tool.outputChars += outputChars;
  tool.outputEvents += 1;
};

export const parseSessionFile = (path: string): SessionAnalysis => {
  const lines = readFileSync(path, "utf8").split(/\r?\n/);
  const events: EventRecord[] = [];
  const tokenEvents: TokenEvent[] = [];
  const toolById = new Map<string, MutableToolCall>();
  let sessionMetaCharsTotal = 0;
  let sessionMetaCharsMax = 0;
  let sessionMetaCount = 0;
  let previousTotalTokens: number | null = null;

  for (const [index, rawLine] of lines.entries()) {
    if (rawLine.length === 0) {
      continue;
    }
    const parsed = parseJsonObject(rawLine);
    if (parsed === null) {
      continue;
    }

    const lineNo = index + 1;
    const ptype = payloadType(parsed);
    const topType = asString(parsed.type);
    const event: EventRecord = {
      lineNo,
      timestamp: asString(parsed.timestamp),
      topType,
      payloadType: ptype,
      rawChars: rawLine.length,
      raw: parsed,
    };
    events.push(event);

    if (topType === "session_meta") {
      sessionMetaCount += 1;
      sessionMetaCharsTotal += rawLine.length;
      sessionMetaCharsMax = Math.max(sessionMetaCharsMax, rawLine.length);
    }

    const payload = isObject(parsed.payload) ? parsed.payload : null;
    if (payload === null) {
      continue;
    }

    if (ptype !== null && CALL_TYPES.has(ptype)) {
      const callId = callIdOf(payload);
      if (callId !== null) {
        const input = decodeInput(payload.arguments ?? payload.input);
        toolById.set(callId, {
          callId,
          timestamp: event.timestamp,
          name: toolNameOf(payload, ptype),
          command: input.label,
          argumentsChars: input.chars,
          callType: ptype,
          inputPreview: input.preview,
          inputChars: input.chars,
          outputChars: 0,
          outputEvents: 0,
        });
      }
    } else if (ptype !== null && OUTPUT_TYPES.has(ptype)) {
      const callId = callIdOf(payload);
      if (callId !== null) {
        const existing = toolById.get(callId);
        if (existing !== undefined) {
          updateOutput(existing, outputCharsOf(payload));
        } else {
          toolById.set(callId, {
            callId,
            timestamp: event.timestamp,
            name: toolNameOf(payload, ptype),
            command: "",
            argumentsChars: 0,
            callType: ptype,
            inputPreview: "",
            inputChars: 0,
            outputChars: outputCharsOf(payload),
            outputEvents: 1,
          });
        }
      }
    } else if (ptype === "token_count") {
      const info = payload.info;
      if (isObject(info)) {
        const total = usageFrom(info.total_token_usage);
        const duplicateTotal = previousTotalTokens === total.totalTokens;
        previousTotalTokens = total.totalTokens;
        tokenEvents.push({
          ...event,
          last: usageFrom(info.last_token_usage),
          total,
          contextWindow: typeof info.model_context_window === "number" ? info.model_context_window : null,
          duplicateTotal,
        });
      }
    }
  }

  const tools = [...toolById.values()].map(toToolCall).sort((a, b) => b.outputChars - a.outputChars);
  const uniqueTokenEvents = tokenEvents.filter((event) => !event.duplicateTotal);

  return {
    path,
    sessionId: basename(path, extname(path)),
    events,
    tokenEvents,
    uniqueTokenEvents,
    tools,
    intervals: buildIntervals(events, tokenEvents, toolById),
    sessionMetaCharsTotal,
    sessionMetaCharsMax,
    sessionMetaCount,
  };
};

const buildIntervals = (
  events: EventRecord[],
  tokenEvents: TokenEvent[],
  toolById: ReadonlyMap<string, MutableToolCall>,
): Interval[] => {
  const byLine = new Map(events.map((event) => [event.lineNo, event]));
  const intervals: Interval[] = [];
  let previousLine: number | null = null;

  for (const token of tokenEvents) {
    if (token.duplicateTotal) {
      continue;
    }

    const start = (previousLine ?? 0) + 1;
    const end = token.lineNo - 1;
    const rawCharsByType = new Map<string, number>();
    const countByType = new Map<string, number>();
    const callIds = new Set<string>();

    for (let line = start; line <= end; line += 1) {
      const event = byLine.get(line);
      if (event === undefined) {
        continue;
      }
      const key = eventTypeKey(event);
      rawCharsByType.set(key, (rawCharsByType.get(key) ?? 0) + event.rawChars);
      countByType.set(key, (countByType.get(key) ?? 0) + 1);
      const payload = event.raw.payload;
      if (isObject(payload)) {
        const callId = callIdOf(payload);
        if (callId !== null) {
          callIds.add(callId);
        }
      }
    }

    const intervalTools = [...callIds]
      .map((callId) => toolById.get(callId))
      .filter((tool): tool is MutableToolCall => tool !== undefined)
      .map(toToolCall)
      .sort((a, b) => b.outputChars - a.outputChars);

    intervals.push({
      previousTokenLine: previousLine,
      tokenEvent: token,
      eventCount: Math.max(0, end - start + 1),
      rawCharsByType,
      countByType,
      tools: intervalTools,
    });
    previousLine = token.lineNo;
  }
  return intervals;
};

const walk = (dir: string): string[] => {
  const files: string[] = [];
  for (const entry of readdirSync(dir)) {
    const path = join(dir, entry);
    const stats = statSync(path);
    if (stats.isDirectory()) {
      files.push(...walk(path));
    } else if (stats.isFile() && path.endsWith(".jsonl")) {
      files.push(path);
    }
  }
  return files;
};

const globToRegExp = (pattern: string): RegExp => {
  const escaped = pattern
    .replace(/[.+^${}()|[\]\\]/g, "\\$&")
    .replace(/\*\*/g, ".*")
    .replace(/\*/g, "[^/]*")
    .replace(/\?/g, ".");
  return new RegExp(`^${escaped}$`);
};

export const discoverFiles = (paths: string[]): string[] => {
  const files = new Set<string>();
  for (const raw of paths) {
    const expanded = raw.replace(/^~/, process.env.HOME ?? "~");
    if (/[?*[]/.test(expanded)) {
      const absolutePattern = resolve(expanded);
      const base = absolutePattern.slice(0, Math.max(absolutePattern.search(/[?*[]/), 1));
      const slash = base.lastIndexOf("/");
      const root = slash > 0 ? base.slice(0, slash) : process.cwd();
      const regex = globToRegExp(absolutePattern);
      if (statSync(root, { throwIfNoEntry: false })?.isDirectory()) {
        for (const file of walk(root)) {
          if (regex.test(resolve(file))) {
            files.add(resolve(file));
          }
        }
      }
      continue;
    }

    const path = resolve(expanded);
    const stats = statSync(path, { throwIfNoEntry: false });
    if (stats?.isDirectory()) {
      for (const file of walk(path)) {
        files.add(resolve(file));
      }
    } else if (stats?.isFile() && path.endsWith(".jsonl")) {
      files.add(path);
    }
  }
  return [...files].sort();
};
